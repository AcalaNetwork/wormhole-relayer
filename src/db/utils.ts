import { isEvmAddress } from '@acala-network/eth-providers';
import { prisma } from './client';

export interface RouterInfo {
  params: string;       // stringified JSON params
  factoryAddr: string;
  recipient: string;
  feeAddr: string;
  routerAddr: string;
}

export const connectDb = async () => {
  try {
    await prisma.$connect();
  } catch (err) {
    console.error('❗️ db connection failed!');
    throw err;
  }
};

const _sanitizeParams = (params: string) => {
  let jsonParams: Record<string, string> = {};
  try {
    jsonParams = JSON.parse(params);
  } catch (err) {
    return params;
  }

  Object.keys(jsonParams).forEach(key => {
    if (isEvmAddress(jsonParams[key])) {
      jsonParams[key] = jsonParams[key].toLowerCase();
    }
  });

  return JSON.stringify(jsonParams);
};

// use lowercase for all addrs to be consistent
export const sanitizeRouterInfo = <T extends Partial<RouterInfo>>(info: T): T => {
  return {
    ...(info.params && { params: _sanitizeParams(info.params) }),
    ...(info.factoryAddr && { factoryAddr: info.factoryAddr.toLowerCase() }),
    ...(info.recipient && { recipient: info.recipient.toLowerCase() }),
    ...(info.feeAddr && { feeAddr: info.feeAddr.toLowerCase() }),
    ...(info.routerAddr && { routerAddr: info.routerAddr.toLowerCase() }),
  } as T;
};
