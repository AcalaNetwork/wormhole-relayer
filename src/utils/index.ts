export * from './configureEnv';
export * from './logger';
export * from './relay';
export * from './route';
export * from './utils';
export * from './validate';
export * from './wormhole';
export * from './address';
export * from './error';
