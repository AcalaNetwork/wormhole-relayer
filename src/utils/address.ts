import { evmToAddr32, nativeToAddr32 } from '@acala-network/asset-router/dist/utils';
import { isEvmAddress, isSubstrateAddress } from '@acala-network/eth-providers';

export const toAddr32 = (addr: string) => {
  if (!isEvmAddress(addr) || !isSubstrateAddress(addr)) {
    throw new Error(`address ${addr} is not a valid evm or substrate address`);
  }

  return isEvmAddress(addr)
    ? evmToAddr32(addr)
    : nativeToAddr32(addr);
};
