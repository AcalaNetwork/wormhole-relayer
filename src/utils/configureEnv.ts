import { ADDRESSES } from '@acala-network/asset-router/dist/consts';
import { CHAIN_ID_ACALA, CHAIN_ID_KARURA, ChainId } from '@certusone/wormhole-sdk';
import dotenv from 'dotenv';

import { ROUTER_CHAIN_ID } from './utils';

dotenv.config({ path: '.env' });

export type RelayerEnvironment = {
  supportedChains: ChainConfig[];
};

export type ChainConfig = {
  chainId: ROUTER_CHAIN_ID;
  ethRpc: string;
  walletPrivateKey: string;
  tokenBridgeAddr: string;
  feeAddr: string;
  factoryAddr: string;
  isTestnet: boolean;
};

const isTestnet = Boolean(Number(process.env.TESTNET_MODE ?? 1));

export function validateEnvironment(): RelayerEnvironment {
  const supportedChains: ChainConfig[] = [];
  supportedChains.push(configKarura());
  supportedChains.push(configAcala());

  return { supportedChains };
}

function configKarura(): ChainConfig {
  const requiredEnvVars = [
    'KARURA_ETH_RPC',
    'KARURA_PRIVATE_KEY',
  ];

  for (const envVar of requiredEnvVars) {
    if (!process.env[envVar]) {
      console.error(`Missing environment variable ${envVar}`);
      process.exit(1);
    }
  }

  const addresses = isTestnet
    ? ADDRESSES.KARURA_TESTNET
    : ADDRESSES.KARURA;

  return {
    chainId: CHAIN_ID_KARURA,
    ethRpc: process.env.KARURA_ETH_RPC!,
    walletPrivateKey: process.env.KARURA_PRIVATE_KEY!,
    isTestnet,
    ...addresses,
  };
}

function configAcala(): ChainConfig {
  const requiredEnvVars = [
    'ACALA_ETH_RPC',
    'ACALA_PRIVATE_KEY',
  ];

  for (const envVar of requiredEnvVars) {
    if (!process.env[envVar]) {
      console.error(`Missing environment variable ${envVar}`);
      process.exit(1);
    }
  }

  const addresses = isTestnet
    ? ADDRESSES.ACALA_TESTNET
    : ADDRESSES.ACALA;

  return {
    chainId: CHAIN_ID_ACALA,
    ethRpc: process.env.ACALA_ETH_RPC!,
    walletPrivateKey: process.env.ACALA_PRIVATE_KEY!,
    isTestnet,
    ...addresses,
  };
}

const env: RelayerEnvironment = validateEnvironment();

export const getChainConfig = (chainId: ChainId): ChainConfig | undefined => (
  env.supportedChains.find((x) => x.chainId === chainId)
);
