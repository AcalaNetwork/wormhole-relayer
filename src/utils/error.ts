import { serialize } from './utils';

export class NoRouteError extends Error {
  constructor(message: string) {
    super(message);
    this.name = 'NoRouteError';
  }
}
export class RelayerError extends Error {
  params: any;
  description: string;

  constructor(message: string, params?: any) {
    super(message);
    this.name = 'RelayerError';
    this.params = params;
    this.description = 'an error occurred!';
  }

  toJson() {
    return {
      msg: this.description,
      error: this.message,
      params: serialize(this.params),
    };
  }
}

export class RelayError extends RelayerError {
  constructor(message: string, params?: any) {
    super(message, params);
    this.name = 'RelayError';
    this.description = 'failed to relay this request!';
  }
};

export class RouteError extends RelayerError {
  constructor(message: string, params?: any) {
    super(message, params);
    this.name = 'RouteError';
    this.description = 'failed to route this request!';
  }
};
