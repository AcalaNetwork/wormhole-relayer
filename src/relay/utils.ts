import {
  hexToUint8Array,
  redeemOnEth,
  parseTransferPayload,
  importCoreWasm,
  ChainId,
  hexToNativeString,
} from '@certusone/wormhole-sdk';
import { ethers } from 'ethers';
import { ChainConfigInfo } from '../configureEnv';
import { EvmRpcProvider } from '@acala-network/eth-providers';
import {
  RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS,
} from './consts';
import axios from 'axios';
import { formatEther } from 'ethers/lib/utils';

interface VaaInfo {
  amount: bigint;
  originAddress: string;
  originChain: ChainId;
  targetAddress: string;
  targetChain: ChainId;
}
interface ShouldRelayResult {
  shouldRelay: boolean;
  msg: string;
} 

export const parseVaa = async (bytes: Uint8Array): Promise<VaaInfo> => {
  const { parse_vaa } = await importCoreWasm();
  const parsedVaa = parse_vaa(bytes);
  const buffered = Buffer.from(new Uint8Array(parsedVaa.payload));

  return parseTransferPayload(buffered);
};

export const shouldRelayVaa = (vaaInfo: VaaInfo): ShouldRelayResult => {
  const {
    amount,
    targetChain,
    originChain,
    originAddress,
  } = vaaInfo;
  const originAsset = hexToNativeString(originAddress, originChain);

  const res = shouldRelay({ targetChain, originAsset, amount });

  console.log('should relay: ', { targetChain, originAsset, amount, res });
  return res;
};

export const shouldRelay = ({
  targetChain,
  originAsset,
  amount: _amount,
}: {
  targetChain: number;
  originAsset: string;
  amount: bigint;
}): ShouldRelayResult => {
  const _noRelay = (msg: string): ShouldRelayResult => ({ shouldRelay: false, msg });

  if (!targetChain) return _noRelay('missing targetChain');
  if (!originAsset) return _noRelay('missing originAsset');
  if (!_amount) return _noRelay('missing transfer amount');

  let amount: bigint;
  try {
    amount = BigInt(_amount);
  } catch (e) {
    return _noRelay(`failed to parse amount: ${_amount}`);
  }

  const supported = RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS[targetChain];
  if (!supported) return _noRelay('target chain not supported');

  const minTransfer = supported[originAsset.toLowerCase()];
  if (!minTransfer) return _noRelay('token not supported');
  if (amount < BigInt(minTransfer)) return _noRelay(`transfer amount too small, expect at least ${minTransfer}`);

  return { shouldRelay: true, msg: '' };
};

export const relayEVM = async (
  chainConfigInfo: ChainConfigInfo,
  signedVAA: string,
  request: any,
  response: any
) => {
  const provider = EvmRpcProvider.from(chainConfigInfo.substrateNodeUrl);
  await provider.isReady();

  const signer = new ethers.Wallet(chainConfigInfo.walletPrivateKey, provider);

  const receipt = await redeemOnEth(
    chainConfigInfo.tokenBridgeAddress,
    signer,
    hexToUint8Array(signedVAA),
  );

  console.log('successfully redeemed on evm', receipt);
  response.status(200).json(receipt);
};

export const fetchBalance = async (address: string, url: string): Promise<number> => {
  try {
    const response = await axios.get(url, {
      data: {
        id: 0,
        jsonrpc: '2.0',
        method: 'eth_getBalance',
        params: [
          address,
          'latest',
        ],
      }
    });

    return Number(formatEther(response.data.result));
  } catch (e) {
    console.log('fetchBalance error: ', e);
    return -1;
  }
};
