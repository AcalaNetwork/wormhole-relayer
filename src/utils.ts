import {
  hexToUint8Array,
  redeemOnEth,
  parseTransferPayload,
  importCoreWasm,
  ChainId,
  hexToNativeString,
  Bridge__factory,
  nativeToHexString,
  CHAIN_ID_ACALA,
  CHAIN_ID_KARURA,
  transferFromEth,
  parseSequenceFromLogEth,
  CHAIN_ID_ETH,
} from '@certusone/wormhole-sdk';
import { BigNumber, BigNumberish, ContractReceipt, ethers, Signer, Wallet } from 'ethers';
import { ChainConfigInfo } from './configureEnv';
import { EvmRpcProvider } from '@acala-network/eth-providers';
import {
  RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS,
} from './consts';
import { parseUnits } from 'ethers/lib/utils';

interface VaaInfo {
  amount: bigint;
  originAddress: string;
  originChain: ChainId;
  targetAddress: string;
  targetChain: ChainId;
}
interface ShouldRelayResult {
  shouldRelay: boolean;
  msg: string;
}

export const parseVaa = async (bytes: Uint8Array): Promise<VaaInfo> => {
  const { parse_vaa } = await importCoreWasm();
  const parsedVaa = parse_vaa(bytes);
  const buffered = Buffer.from(new Uint8Array(parsedVaa.payload));

  return parseTransferPayload(buffered);
};

export const shouldRelayVaa = (vaaInfo: VaaInfo): ShouldRelayResult => {
  const {
    amount,
    targetChain,
    originChain,
    originAddress,
  } = vaaInfo;
  const originAsset = hexToNativeString(originAddress, originChain) as string;  // TODO: update

  const res = shouldRelay({ targetChain, originAsset, amount });

  const info = JSON.stringify({ targetChain, originAsset, amount, res }, (key, value) =>
    typeof value === 'bigint' ? value.toString() : value,
  );
  console.log(`check should relay VAA: ${info}`);

  return res;
};

export const shouldRelay = ({
  targetChain,
  originAsset,
  amount: _amount,
}: {
  targetChain: number;
  originAsset: string;
  amount: bigint;
}): ShouldRelayResult => {
  const _noRelay = (msg: string): ShouldRelayResult => ({ shouldRelay: false, msg });

  if (!targetChain) return _noRelay('missing targetChain');
  if (!originAsset) return _noRelay('missing originAsset');
  if (!_amount) return _noRelay('missing transfer amount');

  let amount: bigint;
  try {
    amount = BigInt(_amount);
  } catch (e) {
    return _noRelay(`failed to parse amount: ${_amount}`);
  }

  const supported = RELAYER_SUPPORTED_ADDRESSES_AND_THRESHOLDS[targetChain];
  if (!supported) return _noRelay('target chain not supported');

  const minTransfer = supported[originAsset.toLowerCase()];
  if (!minTransfer) return _noRelay('token not supported');
  if (amount < BigInt(minTransfer)) return _noRelay(`transfer amount too small, expect at least ${minTransfer}`);

  return { shouldRelay: true, msg: '' };
};

export const relayEVM = async (
  chainConfigInfo: ChainConfigInfo,
  signedVAA: string,
): Promise<ContractReceipt> => {
  const signer = await getSigner(chainConfigInfo);

  const receipt = await redeemOnEth(
    chainConfigInfo.tokenBridgeAddr,
    signer,
    hexToUint8Array(signedVAA),
  );

  await (signer.provider as EvmRpcProvider).disconnect();

  return receipt;
};

export const getSigner = async ({
  nodeUrl,
  walletPrivateKey,
}: {
  nodeUrl: string;
  walletPrivateKey: string;
}): Promise<Signer> => {
  const provider = EvmRpcProvider.from(nodeUrl);
  await provider.isReady();

  return new ethers.Wallet(walletPrivateKey, provider);
};

export const getRouterChainTokenAddr = async (originAddr: string, chainInfo: ChainConfigInfo): Promise<string> => {
  const signer = await getSigner(chainInfo);
  const tokenBridge = Bridge__factory.connect(chainInfo.tokenBridgeAddr, signer);

  return tokenBridge.wrappedAsset(
    CHAIN_ID_ETH,
    Buffer.from(nativeToHexString(originAddr, CHAIN_ID_ETH) as string, 'hex'),   // TODO: use tryNativeToHexString
  );
};

export const bridgeToken = async (
  signer: Wallet,
  tokenBridgeAddr: string,
  coreBridgeAddr: string,
  recipientAddr: string,
  sourceAssetAddr: string,
  targetChain: ChainId,
  amount: BigNumberish,
): Promise<{
  receipt: ContractReceipt;
  sequence: string;
}> => {
  const hexString = nativeToHexString(recipientAddr, targetChain);
  if (!hexString) {
    throw new Error('Invalid recipient');
  }
  const vaaCompatibleAddr = hexToUint8Array(hexString);

  console.log('sending bridging tx...');
  const receipt = await transferFromEth(
    tokenBridgeAddr,
    signer,
    sourceAssetAddr,
    amount,
    targetChain,
    vaaCompatibleAddr,
  );

  const sequence = parseSequenceFromLogEth(receipt, coreBridgeAddr);

  return {
    receipt,
    sequence,
  };
};
