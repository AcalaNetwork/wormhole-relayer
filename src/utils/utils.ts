import { AcalaJsonRpcProvider } from '@acala-network/eth-providers';
import { ApiPromise, Keyring, WsProvider } from '@polkadot/api';
import { BigNumber, Contract, PopulatedTransaction, Signer, Wallet, ethers } from 'ethers';
import { CHAIN_ID_ACALA, CHAIN_ID_AVAX, CHAIN_ID_KARURA, CONTRACTS, hexToUint8Array } from '@certusone/wormhole-sdk';
import { DispatchError } from '@polkadot/types/interfaces';
import { ISubmittableResult } from '@polkadot/types/types';
import { JsonRpcProvider } from '@ethersproject/providers';
import { SubstrateSigner } from '@acala-network/bodhi';
import { options } from '@acala-network/api';
import { parseUnits } from 'ethers/lib/utils';

import { ChainConfig } from './configureEnv';
import { RelayerError } from '../middlewares';
import { SubmittableExtrinsic } from '@polkadot/api/types';
import { bridgeToken, getSignedVAAFromSequence } from './wormhole';

export type ROUTER_CHAIN_ID = typeof CHAIN_ID_KARURA | typeof CHAIN_ID_ACALA;

export const getSigner = async (config: ChainConfig): Promise<Signer> => {
  const provider = new AcalaJsonRpcProvider(config.ethRpc);
  return new ethers.Wallet(config.walletPrivateKey, provider);
};

// TODO: reuse api
export const getApi = async (config: ChainConfig) => {
  const keyring = new Keyring({ type: 'sr25519' });
  const keyPair = keyring.addFromSeed(hexToUint8Array(config.walletPrivateKey));
  const addr = keyPair.address;

  const api = await ApiPromise.create(options({
    provider: new WsProvider(config.nodeUrl),
  }));

  api.setSigner(new SubstrateSigner(api.registry, keyPair));

  return { addr, api };
};

export const parseAmount = async (tokenAddr: string, amount: string, provider: any): Promise<BigNumber> => {
  const erc20 = new Contract(tokenAddr, ['function decimals() view returns (uint8)'], provider);
  const decimals = await erc20.decimals();

  return parseUnits(amount, decimals);
};

export const transferFromAvax = async (
  amount: string,
  sourceAsset: string,
  recipientAddr: string,
  dstChainId: ROUTER_CHAIN_ID,
  wallet: Wallet,
  isMainnet = false,
): Promise<string> => {
  const tokenBridgeAddr = CONTRACTS[isMainnet ? 'MAINNET' : 'TESTNET'].avalanche.token_bridge;
  const coreBridgeAddr = CONTRACTS[isMainnet ? 'MAINNET' : 'TESTNET'].avalanche.core;
  const parsedAmount = await parseAmount(sourceAsset, amount, wallet);
  const { sequence } = await bridgeToken(
    wallet,
    tokenBridgeAddr,
    coreBridgeAddr,
    recipientAddr,
    sourceAsset,
    dstChainId,
    parsedAmount,
  );
  console.log('transfer from AVAX complete', { sequence }, 'waiting for VAA...');

  return getSignedVAAFromSequence(
    sequence,
    CHAIN_ID_AVAX,
    tokenBridgeAddr,
    isMainnet,
  );
};

// TODO: add exitReason to receipt directly
export const getTxFailingReason = async (txHash: string, provider: JsonRpcProvider): Promise<string> => {
  const tx = await provider.getTransaction(txHash);
  try {
    await provider.call(tx as any, tx.blockNumber);
  } catch (err) {
    console.log(err);

    return err.message;
  }

  return '0x';
};

export const getDispatchErrMsg = async (
  api: ApiPromise,
  err: DispatchError,
) => {
  if (err.isModule) {
    const decoded = api.registry.findMetaError(err.asModule);
    const { docs, method, section } = decoded;
    return `${section}.${method}: ${docs.join(' ')}`;
  }

  return err.toString();;
};

export const dryRunExtrinsic = async (
  api: ApiPromise,
  extrinsic: SubmittableExtrinsic<'promise', ISubmittableResult>,
) => {
  // const curBlockHash = (await api.rpc.chain.getBlockHash()).toHex();
  const res = await api.rpc.system.dryRun(extrinsic.toHex());

  if (res.isErr) {
    const err = res.asErr;
    const errMsg = err.toString();

    throw new RelayerError('extrinsic dry run throws validity error', {
      extrinsic: extrinsic.toHex(),
      errMsg,
    });
  }

  if (res.isOk && res.asOk.isErr) {
    const err = res.asOk.asErr;
    const errMsg = getDispatchErrMsg(api, err);

    throw new RelayerError('extrinsic dry run throws dispatch error', {
      extrinsic: extrinsic.toHex(),
      errMsg,
    });
  }
};

export const sendExtrinsic = async (
  api: ApiPromise,
  provider: JsonRpcProvider,
  extrinsic: SubmittableExtrinsic<'promise', ISubmittableResult>,
  confirmations?: number,
  timeout?: number,
) => {
  await dryRunExtrinsic(api, extrinsic);

  return '0x';

  await extrinsic.send();
  const txHash = extrinsic.hash.toHex();
  console.log({ txHash });

  const receipt = await provider.waitForTransaction(txHash, confirmations, timeout);
  if (receipt.status === 0) {
    const reason = await getTxFailingReason(txHash, provider);
    throw new RelayerError('tx failed', { txHash, reason });
  }

  return receipt.transactionHash;
};

export const getEthExtrinsic = async (
  api: ApiPromise,
  provider: JsonRpcProvider,
  tx: PopulatedTransaction,
) => {
  const gasPrice = (await provider.getGasPrice()).toBigInt();
  let gasLimit = 109920n;   // default gas limit

  try {
    gasLimit = (await provider.estimateGas({ ...tx, gasPrice })).toBigInt();
  } catch {
    // use default gas limit
  }

  // const gasPrice = 100009999999n;
  // const gasLimit = 109920n;

  return api.tx.evm.ethCallV2(
    { Call: tx.to },
    tx.data,
    tx.value?.toString(),
    gasPrice,
    gasLimit,
    [],
  );
};

