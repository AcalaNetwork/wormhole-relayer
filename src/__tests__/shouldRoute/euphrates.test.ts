import { describe, expect, it } from 'vitest';

import {
  EUPHRATES_POOLS,
} from '../../consts';
import { RouteParamsEuphrates } from '../../utils';
import {
  expectError,
  shouldRouteEuphrates,
} from '../testUtils';

describe.concurrent.skip('/shouldRouteEuphrates', () => {
  const recipient = '0x0085560b24769dAC4ed057F1B2ae40746AA9aAb6';

  const testShouldRouteEuphrates = async (params: RouteParamsEuphrates) => {
    let res = await shouldRouteEuphrates(params);
    expect(res).toMatchSnapshot();

    // should be case insensitive
    res = await shouldRouteEuphrates({
      ...params,
      recipient: params.recipient.toLocaleLowerCase(),
    });
    expect(res).toMatchSnapshot();
  };

  it('when should route', async () => {
    for (const poolId of EUPHRATES_POOLS) {
      await testShouldRouteEuphrates({
        recipient,
        poolId,
      });
    }
  });

  describe('when should not route', () => {
    it('when missing params', async () => {
      try {
        await shouldRouteEuphrates({
          recipient,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['poolId is a required field'], 400);
      }

      try {
        await shouldRouteEuphrates({
          poolId: 0,
        });
        expect.fail('did not throw an err');
      } catch (err) {
        expectError(err, ['recipient is a required field'], 400);
      }
    });

    it('when bad params', async () => {
      const res = await shouldRouteEuphrates({
        recipient,
        poolId: 520,
      });
      expect(res).toMatchInlineSnapshot(`
        {
          "data": {
            "msg": "euphrates poolId 520 is not supported",
            "shouldRoute": false,
          },
        }
      `);
    });
  });
});
