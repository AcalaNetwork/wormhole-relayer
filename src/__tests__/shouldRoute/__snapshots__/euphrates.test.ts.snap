// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`/shouldRouteEuphrates > when should route 1`] = `
{
  "data": {
    "routerAddr": "0xCc968605e16C9dF27b03aE3E5273507216Fd4C9C",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 2`] = `
{
  "data": {
    "routerAddr": "0xCc968605e16C9dF27b03aE3E5273507216Fd4C9C",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 3`] = `
{
  "data": {
    "routerAddr": "0x84d5Beaa3808834c8A10e8DA241CDc0f1798E546",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 4`] = `
{
  "data": {
    "routerAddr": "0x84d5Beaa3808834c8A10e8DA241CDc0f1798E546",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 5`] = `
{
  "data": {
    "routerAddr": "0xAf808916d378Ee8EE0a737e2bf65Ad4dF1AcF342",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 6`] = `
{
  "data": {
    "routerAddr": "0xAf808916d378Ee8EE0a737e2bf65Ad4dF1AcF342",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 7`] = `
{
  "data": {
    "routerAddr": "0x804Beb16fcd17D3AC84de0614fC9D9e0550BEBf7",
    "shouldRoute": true,
  },
}
`;

exports[`/shouldRouteEuphrates > when should route 8`] = `
{
  "data": {
    "routerAddr": "0x804Beb16fcd17D3AC84de0614fC9D9e0550BEBf7",
    "shouldRoute": true,
  },
}
`;
